### DISCUSSION TITLE
TextLayer - Support for Asian Characters?

### DISCUSSION THREAD

**Discussion by thertzelle** [visgl/deck.gl#5227] [Q&A]
What is the best way to support Asian Character sets in Deck.GL on TextLayer layers? I see that I might have to pass every single Kanji/Katakana/Hiragana character to `characterSet`? What is the best way to do this?

**Comment by Pessimistress** [comment]
You only need to pass the characters that are present in your dataset.

**Comment by thertzelle** [comment]
I'm still trying to wrap my brain around the implementation of this. But potentially that could be 3,400 characters added into this array for that one specific layer. Why doesn't the framework just use the character passed into it to render it?

**Comment by Pessimistress** [comment]
WebGL does not have access to the browser's text rendering capabilities, so TextLayer generates a font atlas image and pass that to the GPU as a texture. It's a trade off whether you want to create a big font that covers all possible characters at the beginning (slower loading) vs only add needed characters as the dataset is evaluated (slower update if the data changes). There are applications that could prefer either. For the use case that this layer was initially tested for - texts in English - creating the font atlas with a small subset of common letters, numbers and symbols is the most efficient, so that became the default.

The layer does not provide an option to automatically generate the character set right now. It is a valid feature we should consider adding. However, if you are concerned about the memory impact of maintaining a character set that has thousands of characters, it will be the same whether it is generated by your app or TextLayer itself.

**Comment by thertzelle** [comment]
Thank you for the explanation, I understand whats going on now. I will close the question.

**Comment by vgrafe** [comment]
Hi @Pessimistress, sorry to necro this chat! 

Same here, I have an app with TextLayers that should display characters from many alphabets. It is a react app, and I use a memoized function that computes the characterset as data changes like the following:

```js
  const characterSet = useMemo(() => {
    const allNames = labels.join(""); // concatenates all labels
    const allCharacters = allNames.split(""); // outputs an array of all characters used
    const dedupedCharacters = Array.from(new Set(allCharacters)); // dedupes the result
    return dedupedCharacters;
  }, [labels]);
```

... and use `characterSet` in the textlayers. Even with a bunch of labels around, the final character count is pretty low, and the `labels` array does not change often, so re-render frequency does not seem to be an issue in my case. 

Does setting `characterSet` to "auto" amounts to a similar logic within the TextLayer? Wondering if adding the logic above is worth the trouble.

I love working with deck.gl, thanks for the efforts on it!
